StereoKit/Framework/Steppers.cs

```diff
@@ -1,74 +1,85 @@
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 
 namespace StereoKit.Framework
 {
 	internal class Steppers
 	{
-		List<IStepper> _steppers = new List<IStepper>();
-
-		public void Shutdown()
+		enum   ActionType { Add, Remove }
+		struct StepperAction
 		{
-			_steppers.ForEach(s => s.Shutdown());
-			_steppers.Clear();
+			public IStepper   stepper;
+			public ActionType type;
+			public StepperAction(IStepper stepper, ActionType type) { this.stepper = stepper; this.type = type; }
 		}
 
+		List           <IStepper>      _steppers = new List<IStepper>();
+		ConcurrentQueue<StepperAction> _actions  = new ConcurrentQueue<StepperAction>();
+
+		// Add steppers via the threadsafe action queue
 		public T Add<T>() where T : IStepper
 		{ 
-			T inst = Activator.CreateInstance<T>(); 
-			Add(inst); 
+			T inst = Activator.CreateInstance<T>();
+			_actions.Enqueue(new StepperAction(inst, ActionType.Add));
 			return inst;
 		}
 		public object Add(Type type)
 		{
 			IStepper inst = Activator.CreateInstance(type) as IStepper;
 			if (inst == null) return null;
-			Add(inst);
+			_actions.Enqueue(new StepperAction(inst, ActionType.Add));
 			return inst;
 		}
 		public T Add<T>(T stepper) where T:IStepper 
 		{
-			// Initialize the stepper! If called from a separate thread, this
-			// needs to happen first, otherwise there's a chance the Step may
-			// happen before Initialize
-			if (SK.IsInitialized)
-				stepper.Initialize();
-
-			// Add the stepper to the list
-			_steppers.Add(stepper);
-
+			_actions.Enqueue(new StepperAction(stepper, ActionType.Add));
 			return stepper;
 		}
-		public void InitializeSteppers()
-		{
-			// This should only be called for steppers that were added before
-			// initialization!
-			System.Diagnostics.Debug.Assert(!SK.IsInitialized);
-			_steppers.ForEach(s => s.Initialize());
-		}
 
+		// Remove steppers via the threadsafe action queue
 		public void Remove<T>() => Remove(typeof(T));
 		public void Remove(Type type)
 		{
-			_steppers.RemoveAll(s => {
-				bool remove = type.IsAssignableFrom(s.GetType());
-				if (remove) s.Shutdown();
-				return remove;
-			});
+			foreach(IStepper stepper in _steppers)
+			{
+				if (type.IsAssignableFrom(stepper.GetType()))
+					_actions.Enqueue(new StepperAction(stepper, ActionType.Remove));
+			}
 		}
 		public void Remove(IStepper stepper)
 		{
-			if (_steppers.Remove(stepper))
-				stepper.Shutdown();
+			_actions.Enqueue(new StepperAction(stepper, ActionType.Remove));
 		}
 
-
 		public void Step()
 		{
-			for (int i = 0; i < _steppers.Count; i++)
+			// Execute all stepper actions on the main thread in the order they
+			// were given.
+			while (_actions.TryDequeue(out StepperAction action))
 			{
-				_steppers[i].Step();
+				switch (action.type)
+				{
+					case ActionType.Add:
+						action.stepper.Initialize();
+						_steppers.Add(action.stepper);
+						break;
+					case ActionType.Remove:
+						_steppers.Remove(action.stepper);
```

Comment:
How about shutting down only if it is found in the list?
(e.g., two threads try to remove a generic type's steppers at ~nearly the same time, so Shutdown isn't called twice for the same stepper)

Reply:
`if (_steppers.Remove(action.stepper)) action.stepper.Shutdown();`

Reply:
Good catch, yeah, probably similar if a stepper's initialization fails.